File: apps/frontend/src/api/rangeInsight.ts
import { apiFetch } from "./client";

// P8 §5/§6 — Range Insight: descriptive daily-range statistics for a symbol.
// Mirrors apps/backend/app/api/v1/schemas/range_insight.py.

export interface MoveStats {
  mean: number;
  median: number;
  p80: number;
}

export interface Band {
  low: number;
  high: number;
}

export interface RangeInsight {
  symbol: string;
  status: "ok" | "insufficient_data";
  bars_used: number;
  low_confidence: boolean;
  as_of: string | null;
  anchor: number | null;
  anchor_source: string | null;
  last_close: number | null;
  atr20: number | null;
  atr20_pct: number | null;
  typical_move_up: MoveStats | null;
  typical_move_down: MoveStats | null;
  support: number | null;
  resistance: number | null;
  high_band: Band | null;
  low_band: Band | null;
  intraday_range: number | null;
  classification: string | null;
  efficiency_ratio: number | null;
  disclaimer: string;
}

// P8 §5a — ranked range-trading candidates (which symbol to range-trade today).
export interface RangeCandidate {
  symbol: string;
  status: string;
  atr20: number | null;
  atr20_pct: number | null;
  intraday_range: number | null;
  classification: string | null;
  last_close: number | null;
  efficiency_ratio: number | null;
  oscillation: number | null;
  suitable: boolean;
  score: number;
  rank: number;
}

export interface RangeCandidatesResponse {
  as_of: string;
  candidates: RangeCandidate[];
}

export const rangeInsightApi = {
  get: (symbol: string) =>
    apiFetch<RangeInsight>(
      `/api/v1/range-insight/${encodeURIComponent(symbol)}`,
    ),
  candidates: (symbols?: string[]) =>
    apiFetch<RangeCandidatesResponse>(
      `/api/v1/range-insight/candidates${
        symbols && symbols.length
          ? `?symbols=${encodeURIComponent(symbols.join(","))}`
          : ""
      }`,
    ),
};

File: apps/backend/app/api/v1/schemas/range_insight.py
"""Pydantic models for ``GET /api/v1/range-insight/{symbol}`` (P8 §5)."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class MoveStatsModel(BaseModel):
    mean: float
    median: float
    p80: float


class BandModel(BaseModel):
    low: float
    high: float


class RangeCandidateModel(BaseModel):
    """One ranked range-trading candidate (P8 §5a)."""

    symbol: str
    status: str
    atr20: float | None
    atr20_pct: float | None
    intraday_range: float | None
    classification: str | None
    last_close: float | None
    efficiency_ratio: float | None
    oscillation: float | None
    suitable: bool
    score: float
    rank: int


class RangeCandidatesResponse(BaseModel):
    """Ranked range-trading candidates over a universe (daily 'what to range-trade' feed)."""

    as_of: datetime
    candidates: list[RangeCandidateModel]


class RangeInsightResponse(BaseModel):
    symbol: str
    status: str  # "ok" | "insufficient_data"
    bars_used: int
    low_confidence: bool
    as_of: datetime | None
    anchor: float | None
    anchor_source: str | None
    last_close: float | None
    atr20: float | None
    atr20_pct: float | None
    typical_move_up: MoveStatsModel | None
    typical_move_down: MoveStatsModel | None
    support: float | None
    resistance: float | None
    high_band: BandModel | None
    low_band: BandModel | None
    intraday_range: float | None
    classification: str | None
    efficiency_ratio: float | None
    disclaimer: str
